tests/test-basic-c.c

```diff
@@ -193,12 +201,67 @@ test_object_writes (gconstpointer data)
     hi_content_stream_new (&hi_memstream, &len, &error);
     g_assert_no_error (error);
     g_autofree guchar *csum = NULL;
-    static const char invalid_hi_sha256[]
-        = "cafebabecafebabecafebabecafebabecafebabecafebabecafebabecafebabe";
     g_assert (!ostree_repo_write_content (repo, invalid_hi_sha256, hi_memstream, len, &csum, NULL,
                                           &error));
     g_assert (error);
     g_assert (strstr (error->message, "Corrupted file object"));
+    g_clear_error (&error);
+  }
+
+  /* Test a basic regfile inline write, no xattrs */
+  g_assert (ostree_repo_write_regfile_inline (repo, hi_sha256, 0, 0, S_IFREG | 0644, NULL,
+                                              (guint8 *)"hi", 2, NULL, &error));
+  g_assert_no_error (error);
+
+  /* Test a basic regfile inline write, erroring on checksum */
+  g_assert (!ostree_repo_write_regfile_inline (repo, invalid_hi_sha256, 0, 0, S_IFREG | 0644, NULL,
+                                               (guint8 *)"hi", 2, NULL, &error));
+  g_assert (error != NULL);
+  g_clear_error (&error);
+
+  static const char expected_sha256_with_xattrs[]
+      = "72473a9e8ace75f89f1504137cfb134feb5372bc1d97e04eb5300e24ad836153";
+
+  GVariantBuilder builder;
+  g_autoptr (GVariant) inorder_xattrs = NULL;
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("a(ayay)"));
+  g_variant_builder_add (&builder, "(^ay^ay)", "security.selinux", "foo_t");
+  g_variant_builder_add (&builder, "(^ay^ay)", "user.blah", "somedata");
+  inorder_xattrs = g_variant_ref_sink (g_variant_builder_end (&builder));
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("a(ayay)"));
+  g_variant_builder_add (&builder, "(^ay^ay)", "user.blah", "somedata");
+  g_variant_builder_add (&builder, "(^ay^ay)", "security.selinux", "foo_t");
+  g_autoptr (GVariant) unsorted_xattrs = g_variant_ref_sink (g_variant_builder_end (&builder));
+
+  /* Now test with xattrs */
+  ostree_repo_write_regfile_inline (repo, expected_sha256_with_xattrs, 0, 0, S_IFREG | 0644,
+                                    inorder_xattrs, (guint8 *)"hi", 2, NULL, &error);
+  g_assert_no_error (error);
+
+  /* And now with a swapped order */
+  ostree_repo_write_regfile_inline (repo, expected_sha256_with_xattrs, 0, 0, S_IFREG | 0644,
+                                    unsorted_xattrs, (guint8 *)"hi", 2, NULL, &error);
+  g_assert_no_error (error);
+
```

Comment:
I notice that you don't have a testcase for what happens if someone tries to repeat an xattr.  If I understand the code above, you'd sort the incoming list, but fail to detect the duplicate, which would later fail to validate.

Do you want to add that check as well?

Reply:
You also don't check for incoming empty xattr key values.  That might be another nice thing to figure out, and then add a test for.

Reply:
Yeah, this normalization also raises the question of what happens in those edge cases. However, this code in the write path isn't returning errors today, doing so would be a notably bigger change.

We don't have active issues with duplicate or zero-key-length xattrs; one would have to explicitly craft things to make that happen today. I just added some sanity checking which is more intended for `fsck` type operations. I think we can leave it as a followup - (really first ensure it's fixed in composefs...which hmm, it looks like we're not sanity checking for empty xattr names there either in a quick look)

Reply:
https://github.com/containers/composefs/pull/400